File: 파이썬/자료구조/queue.py
class Node:
    def __init__(self, data):
        self.data = data
        self.next = None
        
class LinkedList:
    def __init__(self):
        self.head = None
        self.tail = None
        self.size = 0
    
    def append(self, data):
        newNode = Node(data)
        if not self.head:
            self.head = newNode
            self.tail = newNode
            self.size += 1
        else:
            self.tail.next = newNode
            self.tail = newNode
            self.size += 1
    
    def findFirst(self):
        if not self.head:
            return None
        else:
            return self.head.data
    
    def findLast(self):
        if not self.head:
            return None
        else:
            return self.tail.data
            
    def eraseFirst(self):
        if not self.head:
            return None
        else:
            current = self.head
            self.head = current.next
            self.size -= 1
            return current.data
    
    def display(self):
        if not self.head:
            return
        else:
            current = self.head
            answer = []
            while current:
                answer.append(current.data)
                current = current.next
            print(answer, self.size)

class Queue:
    def __init__(self):
        self.list = LinkedList()
            
    def push(self, data):
        self.list.append(data)
        self.list.display()
    
    def pop(self):
        answer = self.list.eraseFirst()
        if not answer:
            print("Queue is empty")
        else:
            print(answer)
            self.list.display()
            return answer
    
    def front(self):
        answer = self.list.findFirst()
        if not answer:
            print("Queue is empty")
        else:
            print(answer)
            return answer
    
    def back(self):
        answer = self.list.findLast()
        if not answer:
            print("Queue is empty")
        else:
            print(answer)
            return answer
        
    def isEmpty(self):
        if self.list.size == 0:
            print("True")
            return True
        else:
            print("False")
            return False
    
    def getSize(self):
        print(self.list.size)
        return self.list.size
    
queue = Queue()
queue.pop()
queue.front()
queue.back()
queue.isEmpty()
queue.getSize()
queue.push(3)
queue.push(5)
queue.pop()
queue.push(10)
queue.push(7)
queue.front()
queue.back()
queue.isEmpty()
queue.getSize()

File: 알고리즘/선형자료구조(Stack,Queue,Dequeue)/src/Queue.java
import java.util.Arrays;

public class Queue {
    private int front;
    private int back;
    private int size;
    private int[] arr;

    public Queue(){
        this.front = 0;
        this.back = 0;
        this.size = 0;
        this.arr = new int[1];
    }

    private void resize(int newCapacity){
        int[] newArray = new int[newCapacity];
        for(int i = 0; i < size; i++){
            newArray[i] = arr[front];
            front = (front + 1) % size;
        }
        this.arr = newArray;
        this.front = 0;
        this.back = size;
        System.out.println("배열 크기 확장됨: " + arr.length);
        System.out.println("front: "+front+" back: "+back);
    }

    public void push(int data){
        if (size == arr.length){
            resize(2*arr.length);
        }
        arr[back] = data;
        back = (back + 1) % arr.length;
        size ++;
        System.out.println(Arrays.toString(arr));
        System.out.println("front: "+front+" back: "+back);
    }

    public int pop(){
        if (size == 0){
            System.out.println("Queue is empty");
            return -1;
        }
        else {
            int pop_var = arr[front];
            front = (front + 1) % arr.length;
            size --;
            System.out.println(Arrays.toString(arr));
            System.out.println("front: "+front+" back: "+back);
            return pop_var;
        }
    }

    public int front(){
        if (size == 0){
            System.out.println("Queue is empty");
            return -1;
        }
        else {
            System.out.println(arr[front]);
            return arr[front];
        }
    }

    public int back(){
        if (size == 0){
            System.out.println("Queue is empty");
            return -1;
        }
        else {
            System.out.println(arr[back-1]);
            return arr[back];
        }
    }

    public boolean isEmpty(){
        if (size == 0){
            System.out.println("True");
            return true;
        }
        else {
            System.out.println("False");
            return false;
        }
    }

    public int getSize(){
        System.out.println(size);
        return size;
    }
}
